pool/src/lib.rs

```diff
@@ -732,45 +747,49 @@ impl LendingPool {
         reserve: &ReserveData,
     ) -> Result<(i128, bool), Error> {
         let debt_token = DebtTokenClient::new(env, &reserve.debt_token_address);
-        let asset_debt = debt_token.balance(who);
+        let borrower_total_debt = debt_token.balance(who);
 
-        if asset_debt == 0 {
+        if borrower_total_debt == 0 {
             return Ok((amount, false));
         }
 
         let underlying_asset = token::Client::new(env, asset);
+        let treasury_address = read_treasury(env);
+
         let debt_coeff = Self::get_debt_coeff(env, reserve)?;
+        let collat_coeff = Self::get_collateral_coeff(env, reserve)?;
 
-        let compounded_debt = debt_coeff
-            .mul_int(asset_debt)
+        let borrower_actual_debt = debt_coeff
+            .mul_int(borrower_total_debt)
             .ok_or(Error::MathOverflowError)?;
 
-        let payback_amount = amount.min(compounded_debt);
+        let borrower_payback_amount = amount.min(borrower_actual_debt);
 
-        let payback_debt = if payback_amount == compounded_debt {
-            asset_debt
-        } else {
-            debt_coeff
-                .recip_mul_int(payback_amount)
-                .ok_or(Error::MathOverflowError)?
-        };
+        let borrower_debt_to_burn = debt_coeff
+            .recip_mul_int(borrower_payback_amount)
+            .ok_or(Error::MathOverflowError)?;
+        let s_token_transfer = collat_coeff
+            .mul_int(borrower_debt_to_burn)
+            .ok_or(Error::MathOverflowError)?;
+        let treasury_amount = borrower_payback_amount
+            .checked_sub(s_token_transfer)
+            .ok_or(Error::MathOverflowError)?;
 
-        underlying_asset.transfer(who, &reserve.s_token_address, &payback_amount);
-        debt_token.burn(who, &payback_debt);
+        underlying_asset.transfer(who, &reserve.s_token_address, &s_token_transfer);
+        underlying_asset.transfer(who, &treasury_address, &treasury_amount);
+        debt_token.burn(who, &borrower_debt_to_burn);
 
         event::repay(env, who.clone(), asset.clone(), amount);
 
-        let remaning_amount = amount
-            .checked_sub(payback_amount)
-            .ok_or(Error::MathOverflowError)?;
-
-        let remaning_amount = if remaning_amount > 0 {
-            remaning_amount
+        let remaning_amount = if amount != i128::MAX && amount > borrower_actual_debt {
```

Comment:
it's not clear that i128::MAX works like `repay_only` flag, may be needs some rework

Reply:
repay_only can be removed, logic with i128::MAX must be described in method's doc